csrc/scheduler/matmul_heuristic_plugin.cpp

```diff
@@ -0,0 +1,210 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <ir/interface_nodes.h>
+#include <mma_type.h>
+#include <scheduler/matmul_heuristic_plugin.h>
+#include <scheduler/mma_utils.h>
+#include <utils.h>
+
+#include <dlfcn.h>
+#include <cstdint>
+#include <mutex>
+
+namespace nvfuser {
+
+namespace matmul_heuristic_plugin {
+
+namespace {
+
+std::mutex plugin_mutex;
+static class PluginInterface : NonCopyable {
+ public:
+  PluginInterface() : filepath_(getNvFuserEnv("MATMUL_HEURISTIC_PLUGIN")) {
+    if (filepath_ != nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      library_handle_ = dlopen(filepath_, RTLD_LAZY);
+      NVF_CHECK(
+          library_handle_ != nullptr,
+          "Error occurred when loading matmul heuristic plugin ",
+          filepath_,
+          ". Error msg: ",
+          dlerror());
+    }
+  }
+
+  bool available() const {
+    return library_handle_ != nullptr;
+  }
+
+  ~PluginInterface() {
+    if (library_handle_ != nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      dlclose(library_handle_);
+      library_handle_ = nullptr;
+      factory_func_ptr_ = nullptr;
+    }
+  }
```

Comment:
Would be nice to unify this with
https://github.com/NVIDIA/Fuser/blob/1110b0aff240e46604283ac4f9fcfe27bc276ff8/csrc/driver_api.cpp#L19-L47

Maybe move both to this file? https://github.com/NVIDIA/Fuser/blob/main/csrc/sys_utils.cpp

Reply:
I took a stab at unifying them into a `LibraryLoader` class in the latest commit.